Editor/EditorCore/DrawShapeTool.cs

```diff
@@ -0,0 +1,375 @@
+        }
+
+        static DrawShapeTool()
+        {
+            s_AvailableShapeTypes = TypeCache.GetTypesDerivedFrom<Shape>();
+        }
+
+        void OnEnable()
+        {
+            InitNewShape();
+            EditorTools.EditorTools.activeToolChanged += ActiveToolChanged;
+            m_ShapeTitle = new GUIContent("Draw Shape");
+            m_ShapeTypesPopupContent = s_AvailableShapeTypes.Select(x => x.Name).ToArray();
+        }
+
+        void InitNewShape()
+        {
+            m_Shape = new GameObject("Shape", typeof(ShapeComponent)).GetComponent<ShapeComponent>();
+            m_Shape.gameObject.hideFlags = HideFlags.HideAndDontSave;
+            m_Shape.hideFlags = HideFlags.None;
+            m_Shape.SetShape(CreateShape(activeShapeType));
+            m_Object = new SerializedObject(m_Shape);
+        }
+
+        void OnDisable()
+        {
+            if (m_Shape.gameObject.hideFlags == HideFlags.HideAndDontSave)
+            {
+                DestroyImmediate(m_Shape.gameObject);
+                InitNewShape();
+            }
+            EditorTools.EditorTools.activeToolChanged -= ActiveToolChanged;
+        }
+
+        Shape CreateShape(Type type)
+        {
+            var shape = Activator.CreateInstance(type) as Shape;
+            ShapeParameters.SetToLastParams(ref shape);
+            return shape;
+        }
+
+        void ActiveToolChanged()
+        {
+            if (EditorTools.EditorTools.IsActiveTool(this))
+                m_InputState = InputState.SelectPlane;
+        }
+
+        void AdvanceInputState()
+        {
+            if (m_InputState == InputState.SelectPlane)
+                m_InputState = InputState.SetWidthAndDepth;
+            else if (m_InputState == InputState.SetWidthAndDepth)
+                m_InputState = InputState.SetHeight;
+            else
+                FinishShape();
+
+            SceneView.RepaintAll();
+        }
+
+        void SetActiveShapeType(Type type)
+        {
+            if (!typeof(Shape).IsAssignableFrom(type))
+                throw new ArgumentException("type must inherit UnityEngine.ProBuilder.Shape", "type");
+
+            s_ActiveShapeIndex.value = s_AvailableShapeTypes.IndexOf(type);
+
+            if (s_ActiveShapeIndex < 0)
+                throw new Exception("type must inherit UnityEngine.ProBuilder.Shape");
+
+            if (m_Shape != null)
+            {
+                if (m_InputState != InputState.SelectPlane)
+                    RebuildShape();
+            }
+        }
+
+        void RebuildShape()
+        {
+            RecalculateBounds();
+
+            if (m_Bounds.size.sqrMagnitude < .01f)
+                return;
+
+            // The sphere doesn't like being built before its height is set
+            if (m_Shape.shape is Sphere && System.Math.Abs(m_Bounds.size.y) < 0.01f)
+                return;
+
+            bool init = false;
+            if (!m_Shape.isInit)
```

Comment:
Is there a special reason to use both init and m_Shape.isInit ? Seems pretty much the same to me here :)

Reply:
I only want to enter the `if(init)` once, when it has just been init. I'll rename to make it more clear :)